### ISSUE TITLE
Collaboration with ember-cli-imgix

### ISSUE THREAD

**Issue by frederickfogerty** [Duder-onomy/ember-imgix#24]
Hey there @Duder-onomy 👋 

Fred from Imgix here. In the past, Imgix has placed less importance on the maintenance of its sdk libraries, due to its startup nature. As a result, some libraries have essentially been unmaintained. This has now changed as Imgix recently hired a full-time SDK Engineer and contracted another SDK Engineer (myself). Our responsibility is to keep these libraries well maintained in the future. 

I am reaching out here to see if there's a possibility for collaboration between our two repos. There are many benefits to collaborating, such as splitting the workload on issue/PR management, and of course development. The two addons seem to have the same public API, with slightly different internals, so it doesn't seem like it would be too hard to work together. 

I'd love to hear your thoughts on collaborating - whether you're interested, and how you'd envisage it would work. A merge would be the best scenario for us.

Hope to hear back from you,
Fred

**Comment by Duder-onomy** [comment]
Hi @frederickfogerty 
I would totally love to work together. 
The only reason I created this lib is that I was waiting over 1 year for a bug fix PR to get merged into ember-cli-imgix.

Having the canonical ember + imgix solution makes more sense in the imgix namespace. 
My only fear of joining back up is that if we cannot get PR's through anymore (like happened previous). We use imgix at thedyrt.com for 95% of our images. And we use ember. SO, this integration point (and the ability to fix things / add features ) is very important for us. 

I don't think a straight up merge will work as I have refactored / completely re-written allot of things and I think those merge conflicts would be pretty brutal. 
My preferred path would be to get a short list of things from you and your team that you would like to see updated or changed in this repo in order for you to feel good about it being a 'supported' imgix solution. I can make those changes and updates. When you feel good about it, we can replace the current ember-cli-imgix source with the source of this repo. I am down to talk through alternative pathways as well. 

My short list of things I had envisioned adding in the next few weeks / days:
* Better (opt in only) fastboot support with LQIP. (working on this as I type... we need it for a new feature at TheDyrt)
* Try and remove the `fastboot-transform` we are wrapping around the imgix client sdk to avoid runtime errors in node... I think the only bit of work there is to use a diff URL parsing lib when running in a node env vs browser.
* Move explicit imgix api options from the top level component options into the options hash.
* Better documentation and interactive examples (maybe use ember-cli-addon-docs) 

Anyhow. Happy to chat about it more here, or you can email me greg@thedyrt.com

**Comment by frederickfogerty** [comment]
That's great to hear @Duder-onomy! I can imagine how frustrating that must have been not to have a PR merged for a year when your business depends on it, so I must apologise for that. 

The importance of this library to your business is evident, so it is crucial that you remain able to make changes to the library. It seems that this would be solved by adding you as a collaborator of `ember-cli-imgix` - then you would be able to merge PRs as you see fit. With our other projects, we require the status checks and at least 1 PR review before the PR can be merged. This is to uphold the quality of the repo. I see this being your only barrier to merging a PR, but at the moment we've been prompt to review these and merge. You would also be able to have a colleague review your PRs if we were unavailable. Does this arrangement fit your needs?

I agree that a straight `git pull ember-imgix:master master` is not going to work due to the significant changes. My meaning behind "merge" was more to take the best parts from both repos and combine them.

That plan overall sounds good to me. Creating an exhaustive shortlist will require looking at these repositories side-by-side in detail. As we're switching our focus over to our React library for a little while, we won't be able to produce a complete shortlist for 1-2 weeks. Overall `ember-imgix` looks to be in pretty good shape, so I don't imagine it will be much.

Without stepping into the code I can say we'll need a few things before we can merge:

- Compatibility with our current API (as much as possible). Although this doesn't need to be a complete match (as this update will be a breaking change), for the users of `ember-cli-imgix` it will be easier to migrate to the new version if the APIs are similar. For example, I think we'll need to include support for `imgix-image-element`. If it's not useful or not being used I'm happy to deprecate it and remove it in the next major version, but I don't think we can remove it straight away.

- A priority for us is comprehensive unit and integration tests. As I can see that the tests for this library are quite good, I doubt that we'll need to add many more tests, if any at all. I simply want to highlight that this is important to us and something we'll need to check off before consolidating.

I appreciate the personal contact but I decided to keep this discussion open-source, as, well, we are talking about open-source software after all.

**Comment by Duder-onomy** [comment]
Awesome. Everything sounds great. 

I will spend some time this weekend setting up a simple ember app that uses both the libs side-by-side.  This will make it easier for me to compile a list of changes + a list of things we will want to bring back (if only to deprecate) 

I will keep in touch throughout the process.

**Comment by frederickfogerty** [comment]
Thank you so much @Duder-onomy. Let me know if you need anything from me during the process.

**Comment by frederickfogerty** [comment]
Hey @Duder-onomy just wanted to check in to see how things are going, and if I could help with anything? I have some time set aside this week to get the ball rolling on this, so let me know!

**Comment by frederickfogerty** [comment]
Hey there @Duder-onomy I've started the merge on [this branch](https://github.com/imgix/ember-cli-imgix/tree/fred/merge-ember-imgix).

Previously we had `imgix-image` and `imgix-image-element`. I've remove `imgix-image-element` and renamed the existing `imgix-image` to `imgix-image-wrapper`. I've then taken the source from this repo's `imgix-image` and used that for the `imgix-image component`. In summary: 

```
imgix-image - replaced by ember-imgix/imgix-image
imgix-image-element - deleted
imgix-image-wrapper - renamed from imgix-image 
```

I think it's a good idea to keep both components for the next major version so that the migration is easier. `imgix-image-wrapper` should be deprecated in favour of `imgix-image`, and will be removed in the following major version.

I had a question - what is `/tests/dummy/config/targets.js` used for? I'm new to the latest versions of ember, and just wondered if this was used anywhere.

The next steps are to ensure that all of the tests that covered the old `imgix-image` still apply to the new `imgix-image` and `imgix-image-wrapper`

**Comment by frederickfogerty** [comment]
Hey @Duder-onomy I've completed my work in merging the two repos, and I've submitted a PR on `ember-cli-imgix` [here](https://github.com/imgix/ember-cli-imgix/pull/31). It would be a great if you could give this a review and let me know what you think.

**Comment by Duder-onomy** [comment]
@frederickfogerty Great! I was just sitting down to jump into this. You have done most of it. Ill put my comments over there!

**Comment by frederickfogerty** [comment]
@Duder-onomy I look forward to your comments. We can set you up as a contributor over there next week as well.

**Comment by frederickfogerty** [comment]
Hey @Duder-onomy did you have a chance to look at that PR? We'd love to get that merged in the next few days 🙌

**Comment by Duder-onomy** [comment]
Cool. This is done. Going to deprecate ember-imgix

**Comment by frederickfogerty** [comment]
Thanks for all your help @Duder-onomy 🙌Glad we could get these libraries merged
